File: core/lua/jitaccess/core/pae.lua
-- jitaccess.core.pae — Pre-Authentication Encoding (PASETO-compatible)
--
-- LE64(n)    = 8-byte little-endian of n, top bit of the last byte cleared
-- PAE(parts) = LE64(#parts) .. for each p: LE64(#p) .. p
--
-- Makes a list of variable-length byte strings injective, so no two distinct
-- field lists collide inside an HMAC (SECURITY-REVIEW H1). See core/SPEC.md §2,
-- docs/PROTOCOL.md §5.1. Pinned by core/testdata/vectors.json ("pae").
--
-- NOTE: not executed on the dev host (no Lua runtime); validated against the
-- shared vectors by the docker harness. Pure module — no ngx/resty deps.

local concat = table.concat
local char = string.char
local floor = math.floor

local _M = { _VERSION = "0.1.0" }

-- 8-byte little-endian, high bit of the final byte cleared (PASETO PAE).
-- Lengths here are tiny; the div/mod loop avoids any 32/64-bit bitop width traps.
local function le64(n)
  local b = {}
  for i = 1, 8 do
    b[i] = char(n % 256)
    n = floor(n / 256)
  end
  -- clear the top bit of the last byte without needing the bit library
  b[8] = char(string.byte(b[8]) % 128)
  return concat(b)
end
_M.le64 = le64

-- parts: array-like table of byte strings
function _M.encode(parts)
  local out = { le64(#parts) }
  local j = 1
  for i = 1, #parts do
    local p = parts[i]
    j = j + 1; out[j] = le64(#p)
    j = j + 1; out[j] = p
  end
  return concat(out)
end

return _M

File: core/go/pae.go
// Package jitcore is the Go reference implementation of the JIT Network Access
// portable core (L3). It is the library the standalone Authorizer and the native
// Caddy module embed, and it is the second implementation (after core/lua) that
// must agree byte-for-byte with core/testdata/vectors.json.
//
// See ../SPEC.md for the interface contract and ../../docs/PROTOCOL.md for the
// wire format. The pure parts of this package (PAE, canonicalization, proof and
// nonce crypto) are fully pinned by the shared vectors; vectors_test.go asserts
// every one of them. The only stateful parts are GrantStore and NonceStore.
package jitcore

import "encoding/binary"

// LE64 is the PASETO length encoding: 8-byte little-endian with the top bit of
// the final byte cleared.
func LE64(n uint64) []byte {
	b := make([]byte, 8)
	binary.LittleEndian.PutUint64(b, n)
	b[7] &= 0x7f
	return b
}

// PAE is Pre-Authentication Encoding (PROTOCOL §5.1).
//
// Concatenating variable-length fields before a MAC is not injective: distinct
// field lists can produce identical bytes, which allowed a cross-service proof
// collision in the original draft (SECURITY-REVIEW H1). Length-prefixing every
// piece — and the piece count — makes the encoding injective, so a proof minted
// for one (service, kid, nonce) can never be reinterpreted as another.
func PAE(pieces ...[]byte) []byte {
	n := 8
	for _, p := range pieces {
		n += 8 + len(p)
	}
	out := make([]byte, 0, n)
	out = append(out, LE64(uint64(len(pieces)))...)
	for _, p := range pieces {
		out = append(out, LE64(uint64(len(p)))...)
		out = append(out, p...)
	}
	return out
}
